### PR TITLE
Add named pipe support for windows

### PR DESCRIPTION
**Pull request by randomanderson**
This PR adds named pipe support for windows machines. Named pipes are similar to unix domain sockets. This PR supports submitting metrics to Datadog in Azure App Services (https://github.com/DataDog/datadog-aas-extension).

The main changes are:
* Abstracting `DatagramChannel` into the `ClientChannel` interface
* Adding a `ClientChannel` implementation for named pipes
* Adding configuration support in the builder
* Subclassing `DummyStatsD` server for testing of the 3 transport types

Unfortunately, only *existing* named pipes can be written to in pure Java. In order to create named pipes for testing, JNA is used.

### CODE REVIEW

**Review by vickenty — COMMENTED** [review]
Looks mostly good. Left a few comments inline, please let me know what you think.

File: src/main/java/com/timgroup/statsd/ClientChannel.java [hunk 1]
```diff
@@ -0,0 +1,10 @@
+package com.timgroup.statsd;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.nio.ByteBuffer;
```
[the remark this answers is not in the document]
Comment (randomanderson) [reply src/main/java/com/timgroup/statsd/ClientChannel.java]: fixed

File: src/main/java/com/timgroup/statsd/NamedPipeClientChannel.java [hunk 2]
```diff
@@ -0,0 +1,39 @@
+package com.timgroup.statsd;
+
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.io.RandomAccessFile;
+import java.nio.ByteBuffer;
+import java.nio.channels.FileChannel;
+import java.nio.channels.WritableByteChannel;
```
[the remark this answers is not in the document]
Comment (randomanderson) [reply src/main/java/com/timgroup/statsd/NamedPipeClientChannel.java]: fixed

File: src/main/java/com/timgroup/statsd/NonBlockingStatsDClient.java [hunk 3]
```diff
@@ -427,6 +396,14 @@ public void stop() {
                     handler.handle(e);
                 }
             }
+
+            if (telemetryClientChannel != null) {
```
[the remark this answers is not in the document]
Comment (randomanderson) [reply src/main/java/com/timgroup/statsd/NonBlockingStatsDClient.java]: Yes. I added a check even though `close()` should be idempotent.

File: src/main/java/com/timgroup/statsd/NamedPipeSocketAddress.java [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+    }
+
+    public String getPipe() {
+        return pipe;
+    }
+
+    /**
+     * A normalized version of the pipe name that includes the `\\.\pipe\` prefix
+     */
+    public static String normalizePipeName(String pipeName) {
```
[the remark this answers is not in the document]
Comment (randomanderson) [reply src/main/java/com/timgroup/statsd/NamedPipeSocketAddress.java]: This method is used in the tests so I made it package-private

File: src/main/java/com/timgroup/statsd/NonBlockingStatsDClient.java [hunk 5]
```diff
@@ -469,6 +446,26 @@ StringBuilder tagString(final String[] tags, StringBuilder builder) {
...
+            UnixDatagramChannel unixDatagramChannel = UnixDatagramChannel.open();
+            // Set send timeout, to handle the case where the transmission buffer is full
+            // If no timeout is set, the send becomes blocking
+            if (timeout > 0) {
+                unixDatagramChannel.setOption(UnixSocketOptions.SO_SNDTIMEO, timeout);
+            }
+            if (bufferSize > 0) {
+                unixDatagramChannel.setOption(UnixSocketOptions.SO_SNDBUF, bufferSize);
+            }
+            return new DatagramClientChannel(unixDatagramChannel, address);
```
[the remark this answers is not in the document]
Comment (randomanderson) [reply src/main/java/com/timgroup/statsd/NonBlockingStatsDClient.java]: done.

File: src/main/java/com/timgroup/statsd/StatsDSender.java [hunk 6]
```diff
@@ -92,14 +89,13 @@ void sendLoop() {
...
 
                 buffer.flip();
-                final int sentBytes = clientChannel.send(buffer, address);
+                final int sentBytes = clientChannel.write(buffer);
 
                 buffer.clear();
                 if (sizeOfBuffer != sentBytes) {
                     throw new IOException(
-                            String.format("Could not send stat %s entirely to %s. Only sent %d out of %d bytes",
+                            String.format("Could not send stat %s entirely. Only sent %d out of %d bytes",
```
[the remark this answers is not in the document]
Comment (randomanderson) [reply src/main/java/com/timgroup/statsd/StatsDSender.java]: done

File: src/main/java/com/timgroup/statsd/NonBlockingStatsDClientBuilder.java [hunk 7]
```diff
@@ -193,7 +199,11 @@ NonBlockingStatsDClientBuilder resolve() {
         Callable<SocketAddress> telemetryLookup = telemetryAddressLookup;
 
         if (lookup == null) {
-            lookup = staticStatsDAddressResolution(hostname, port);
+            if (namedPipe == null) {
```
[the remark this answers is not in the document]
Comment (randomanderson) [reply src/main/java/com/timgroup/statsd/NonBlockingStatsDClientBuilder.java]: This was added. It uses the same env variable as the AAS extension

**Comment by randomanderson** [comment]
@vickenty Did you want to rereview or should I just merge this PR with the fixes?

**Comment by vickenty** [comment]
@randomanderson Sorry for the delay, I will take another look. Meanwhile, we would like to make sure configuration options are the same between libraries for different languages. Windows pipes are not universally supported by all libraries we support, but where it is available, we'd like the configuration to be the same (same env vars / methods). Can you please check if we're aligned with https://github.com/DataDog/datadog-go for example?

**Comment by randomanderson** [comment]
@vickenty Windows pipes are only implemented in .NET and the dogstatsd server. They both use `DD_DOGSTATSD_PIPE_NAME` which is what I used here

**Comment by vickenty** [comment]
There are some warnings in the CI, would you mind taking a look and fixing if any are new?

Do you think it might be useful to throw an error if named pipe is requested on a non-windows OS?

File: src/main/java/com/timgroup/statsd/NonBlockingStatsDClient.java [hunk 8]
```diff
@@ -469,6 +445,17 @@ StringBuilder tagString(final String[] tags, StringBuilder builder) {
...
     }
 
+    ClientChannel createByteChannel(Callable<SocketAddress> addressLookup, int timeout, int bufferSize) throws Exception {
+        final SocketAddress address = addressLookup.call();
+        if (address instanceof NamedPipeSocketAddress) {
+            return new NamedPipeClientChannel((NamedPipeSocketAddress) address);
+        } else if (address instanceof UnixSocketAddress) {
+            return new UnixDatagramClientChannel(address, timeout, bufferSize);
+        } else {
+            return new DatagramClientChannel(DatagramChannel.open(), address);
```
Comment (vickenty) [inline src/main/java/com/timgroup/statsd/NonBlockingStatsDClient.java:455]: Would this work?
```suggestion
            return new DatagramClientChannel(address);
```
And then DatagramChannel import can be removed, and maybe 2-argument constructor too (other client channel impls don't have one, so I guess we don't need one there too).

**Review by vickenty — APPROVED** [review]
LGTM, except one small nit.

**Comment by randomanderson** [comment]
@vickenty I made that small change. I left the 2-arg constructor because `UnixDatagramClientChannel` uses it.

**Pull request merged**